### PR TITLE
Ruby on Rails capstone project - Budget App: Release 1.0.0

### PR DESCRIPTION
**Pull request by selma-belhadj**
## What does this PR do?
This PR is raised to  build a mobile web application where the user can manage his budget: he has a list of transactions associated with a category, so that he can see how much money he spent and on what.
- [Live Demo](https://afternoon-anchorage-61931.herokuapp.com/)
- [Video presentation](https://loom.com/share/862f7dc6b65b4721ac0f)

## Description of the task to be completed
- Interactions
- Added Splash screen
  - A simple page with the name 'Cash App' and links to the sign up and log in pages.
- Added Sign up and log in pages
  - The user can register in the app with full name, email and password (all mandatory).
  - The user can log into the app using email and password.
- Added Home page (categories page) When the user logs in, he is presented with the categories page. 
  - For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
  - When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
  - There is a button "add a new category" at the bottom that brings the user to the page to create a new category.
- Added Transactions page
  - For a given category, the list of transactions is presented, ordered by the most recent.
  - At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
  - There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
  When the user clicks on the "Back" button (<), the user navigates to the home page.
- Added a new category page
  - The user fills out a form to create a new category, indicating their name and icon (both mandatory).
  - The user click (or taps) the "Save" button to create the new category, and is taken to the home page on success.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.
- Added a new transaction page
  - The user fills out a form to create a new transaction with: name, amount  & categories 
  - The user click (or taps) the "Save" button to create the new category, and is taken to the transactions page for that category.
  - When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.
- Testing requirements
  - Created unit and integration tests for all the most important components of the RoR application.
 
## How should it be tested

- clone the repository
  `git clone https://github.com/selma-belhadj/BudgetApp.git`
- navigate to the folder
  `cd BudgetApp`
- checkout the feature
  `git checkout release_1.0.0`
- Install packages
  `bundle install`
- Setup database
  `rails db:setup`
- Run migration
  `rails db:migrate`

### CODE REVIEW

File: app/views/categories/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<header>
```
Comment (arslanbisharat) [inline app/views/categories/index.html.erb:1]: - [x] When I try to save a category with no URL or invalid URL, I get the following error. Please make sure you have custom error for this. 

image

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  #
+  #   # Run specs in random order to surface order dependencies. If you find an
+  #   # order dependency and want to debug it, you can fix the order by providing
+  #   # the seed, which is printed after each run.
+  #   #     --seed 1234
+  #   config.order = :random
+  #
+  #   # Seed global randomization in this process using the `--seed` CLI option.
+  #   # Setting this allows you to use `--seed` to deterministically reproduce
+  #   # test failures related to randomization by passing the same `--seed` value
```
Comment (arslanbisharat) [inline spec/spec_helper.rb:89]: - [x] Please make sure App has unit tests for all the most important components of the RoR application.

File: .github/workflows/linters.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
+name: Linters
```
Comment (arslanbisharat) [inline .github/workflows/linters.yml:1]: - [x] You have mutiple .DS Store files. Please add them to gitignore to make your PR clean.

File: config/routes.rb [hunk 4]
```diff
@@ -1,5 +1,11 @@
 Rails.application.routes.draw do
```
Comment (arslanbisharat) [inline config/routes.rb:1]: - [x] The routes are not present correcltu. Please make sure you have followed the correct rotues so it leads to correct page.

File: README.md [hunk 5]
```diff
@@ -1,24 +1,61 @@
...
 
-* Ruby version
+* register and log in, so that the data is private to them.
+* introduce new transactions associated with a category.
+* see the money spent on each category.
 
-* System dependencies
+## [Live Demo](https://afternoon-anchorage-61931.herokuapp.com/)
 
-* Configuration
```
Comment (arslanbisharat) [inline README.md:12]: - [x] In recorded video please address a piece of code/feature you liked about the project.

**Review by arslanbisharat — CHANGES_REQUESTED** [review]
Hi @selma-belhadj, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

- The Pull Request has a proper title and description. ✔️
- You have added a descriptive Readme file. Good Job. ✔️
- There is no linter error is present. ✔️

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- [ ] Please make sure you right descriptive comments. Commenting involves placing Human Readable Descriptions inside of computer programs detailing what the Code is doing. Proper use of commenting can make code maintenance much easier, as well as helping make finding bugs faster. Further, commenting is very important when writing functions that other people will use. :x:


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/categories/index.html.erb (hunk 1 above)
Comment (selma-belhadj) [reply app/views/categories/index.html.erb:1]: Hello @arslanbisharat, I have just tried the same case and got no error. 
Screenshot 2022-08-06 at 14 19 55

File: spec/spec_helper.rb (hunk 2 above)
Comment (selma-belhadj) [reply spec/spec_helper.rb:89]: Hello @arslanbisharat  again, Here are the tests I have implemented and the results:

Screenshot 2022-08-06 at 14 29 16
Screenshot 2022-08-06 at 14 29 32

File: config/routes.rb (hunk 4 above)
Comment (selma-belhadj) [reply config/routes.rb:1]: The app works correctly and the routes are well defined: each transaction belongs to one category

File: README.md (hunk 5 above)
Comment (selma-belhadj) [reply README.md:12]: The delete option for the transaction and category which is not mentioned in the requirements but I did it

File: .github/workflows/linters.yml (hunk 3 above)
Comment (selma-belhadj) [reply .github/workflows/linters.yml:1]: Hello @arslanbisharat , I think this task is not required but I have added them as mentioned.

**Review by ahmetbozaci — APPROVED** [review]
# 🥳 🎉🥇 💯  💥 🎆 🎈 STATUS: APPROVED 🥳 🎉  🥇  💯  💥   🎆 🎈 🎉 



@selma-belhadj  ✋🏼 
#### Your project is great. 🎉 I approve of it. There is nothing else to say other than... it's time to merge it. You can submit this project on your dashboard.
___ 

### 💯 Highlights 💯
- Your git-flow is correct. ✔️
- You set up linter files and no errors. ✔️
- Your pull request is very professional. It has a meaningful title and summary. Keep it up.✔️
- You create a presentation video to a demo of the website and highlight the code that you are proud of it.✔️
- You follow the design guidelines(color, layout, typography)✔️
- You create tests.✔️
- I can see the list of transactions ordered by the most recent and ı can see the total amount on the top.✔️

___ 

### Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
@selma-belhadj @selma-belhadj 
- [ ] Pull request flow is wrong. You should open PR development to the feature branch. [Requirement](https://github.com/microverseinc/curriculum-rails/blob/main/capstone/rails_capstone.md#:~:text=Make%20sure%20that%20you%20used%20correct%20gitflow.)
image

Example for git-flow
image


⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ 

- Your commit messages should always start with capital letters and they should be an imperative tense. I see some of them are not like that. Instead of writing `Added style or Created the file` you should write `Add style or Create the file`
- ***I write it as optional. But this is a [general requirement](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md#:~:text=The%20commit%20message%20should%20be%20written%20in%20the%20imperative%20tense%20and%20capitalized) for every project. So you should pay attention to this in your next projects.***
- You can change them by using `git-rebase`. If you want to know how to change ı leave a few resources.
- [Video - How to change commit messages](https://www.loom.com/share/00771f369b7b47fc968c019ecf0d3ba6)
- [Article - How to create good commit messages](https://www.freecodecamp.org/news/level-you-up-to-awesome-commit-messages-a85558cb90e8/)

⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ 
___

### Cheers and Happy coding! 💻 👏
aproved

___

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification._
___

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form.](https://airtable.com/shrQAqnBwek5a0O0s)_

[GitHub](https://github.com/ahmetbozaci) | [LinkedIn](https://linkedin.com/in/ahmetbozaci) | [Twitter](https://twitter.com/ahmtbozaci)

_Reviewed by [Ahmet Bozacı ](https://github.com/ahmetbozaci)_

**Pull request merged**